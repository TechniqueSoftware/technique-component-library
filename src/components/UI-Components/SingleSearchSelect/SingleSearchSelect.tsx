import * as React from 'react';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { createStyles, Theme } from '@material-ui/core';
import withStyles from '@material-ui/styles/withStyles';
import Fade from '@material-ui/core/Fade';
import Grid from '@material-ui/core/Grid';
import Search from '@material-ui/icons/Search';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';

import TextInput, { TextInputProps } from '../TextInput/TextInput';
import { SelectCustom, SelectCustomProps } from '../Select/SelectCustom';

export interface SingleSearchSelectProps extends SelectCustomProps {
  options: {
    label: string,
    value: string
  }[];
  emptyDataLabel?: string;
  searchContainerMaxWidth?: number;
  listContainerHeight?: number;
  error?: boolean;
  classes?: {
    container?: string;
  };
  autoFocus?: boolean;
  placeholder?: string;
  ariaLabel?: string;
  startAdornmentIcon?: React.ReactElement;
  onChange: (value: any) => void;
}

const useSingleSearchSelectPropsStyles = makeStyles(() =>
  createStyles({
    searchContainer: (props: SingleSearchSelectProps) => {
      return ({ width: props.searchContainerMaxWidth || 400 });
    },
    container: {},
  }));

const styles = (theme: Theme) => ({
  container: {
    border: `1px solid ${theme.palette.grey['300']}`,
    background: theme.palette.common.white
  }
});

function SingleSearchSelect(props: SingleSearchSelectProps) {
  const {
    value,
    options = [],
    maxWidth = 300,
    id,
    label,
    disabled = false,
    open,
    emptyDataLabel = '',
    error,
    listContainerHeight,
    hideLabel = false,
    classes,
    autoFocus = false,
    placeholder = 'Search',
    ariaLabel = 'Search',
    startAdornmentIcon = <Search />,
    onChange,
    onMenuOpen,
    onMenuClose
  } = props;

  const [search, setSearch] = React.useState<string>('');

  const handleOpenMenu = () => {
    onMenuOpen();
  };

  const handleCancel = () => {
    onMenuClose();
    setSearch('');
  };

  const handleSearchChange = (event: any) => {
    const search = event.target.value;
    setSearch(search);
  };

  const handleSelect = (value: string) => {
    onChange(value);
    handleCancel();
  };

  const renderListItems = () => {
    const listStyles = { overflow: 'auto', maxHeight: listContainerHeight, width: '100%' };
    const options = getFilteredOptions();

    return (
      <List style={listStyles} dense>
        {
          options.map(option => (
            <ListItem
              dense
              key={option.value}
              selected={option.value === value}
              style={{ cursor: 'pointer' }}
              onClick={() => handleSelect(option.value)}
            >
              <Typography component="span">
                {option.label}
              </Typography>
            </ListItem>
          ))
        }
        {(search && !options.length) && <ListItem><ListItemText>No match found</ListItemText></ListItem>}
      </List>
    );
  };

  const getFormattedValue = (value: string) => {
    const option = options.find(x => x.value === value);
    return option ? option.label : !options.length ? emptyDataLabel : '';
  };

  const getFilteredOptions = () => {
    return options.filter(option => option.label.toLowerCase().includes(search.toLowerCase()));
  };

  const selectCustomProps: SelectCustomProps = {
    disabled,
    label,
    hideLabel,
    maxWidth,
    open,
    value: getFormattedValue(value),
    id: `${id}-single-search-select`,
    onMenuOpen: handleOpenMenu,
    onMenuClose: handleCancel
  };

  const textInputProps: TextInputProps = {
    error,
    autoFocus,
    placeholder,
    startAdornmentIcon,
    value: search,
    'aria-label': ariaLabel,
    onChange: handleSearchChange,
  };

  const { searchContainer } = useSingleSearchSelectPropsStyles(props);

  return (
    <SelectCustom {...selectCustomProps} >
      <Fade in={true}>
        <Grid item className={searchContainer}>
          <div className={classes.container}>
            <TextInput {...textInputProps} />
            {renderListItems()}
          </div>
        </Grid>
      </Fade>
    </SelectCustom>
  );
}

export default withStyles(styles)(SingleSearchSelect);
