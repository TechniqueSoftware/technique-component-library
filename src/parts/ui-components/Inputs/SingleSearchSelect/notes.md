# SingleSearchSelect
This component wraps our `SelectCustom` component. Had similar behavior like `MultiSelectFieldGroupButton`, but is to be used for select one value

## Usage
This component was created initially to be used in the `UserSearch` page

## JIRA
Related jira ticket: https://clubos.atlassian.net/browse/HELP-2250
