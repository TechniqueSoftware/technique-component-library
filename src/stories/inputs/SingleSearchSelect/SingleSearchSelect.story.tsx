import * as React from 'react';
import { createCUICStoriesOf } from '../../utils/utils';
import { MODULE_NAME } from '../../constants';
import { clubOsTheme, SingleSearchSelect, SingleSearchSelectProps } from '@clubos-ui-components/core';
import ThemeProvider from '@material-ui/styles/ThemeProvider';
import Box from '@material-ui/core/Box';
import { boolean, text, number } from '@storybook/addon-knobs';

require('moment-timezone');

function SingleSearchSelectStory() {
  const [open, setOpen] = React.useState(false);
  const [value, setValue] = React.useState('');
  const listContainerHeight = number('List container height', 450);
  const maxWidth = number('Select max width', 300);
  const searchContainerMaxWidth = number('Search container max width', 400);
  const label = text('Label', 'Nouns');
  const hideLabel = boolean('Hide label', false);
  const disabled = boolean('Disabled', false);
  const emptyDataLabel = text('Empty data label', '');
  const error = boolean('Has error', false);
  const placeholder = text('Search placeholder', 'Search');
  const autoFocus = boolean('Autofocus', false);

  const props: SingleSearchSelectProps = {
    open,
    value,
    listContainerHeight,
    maxWidth,
    searchContainerMaxWidth,
    label,
    disabled,
    hideLabel,
    emptyDataLabel,
    error,
    placeholder,
    autoFocus,
    id: 'selector',
    options: emptyDataLabel ? [] : [
      { label: 'All', value: '' },
      { label: 'test 1', value: '1' },
      { label: 'test 2', value: '2' },
      { label: 'test 3', value: '3' }
    ],
    onChange: (value: React.SetStateAction<string>) => setValue(value),
    onMenuOpen: () => setOpen(true),
    onMenuClose: () => setOpen(false)
  };

  return (
    <ThemeProvider theme={clubOsTheme}>
      <Box p={4}>
        <SingleSearchSelect {...props} />
      </Box>
    </ThemeProvider>
  );
}

createCUICStoriesOf(MODULE_NAME.INPUTS, module)
  .addCUICStory({
    render: () => {
      return (<SingleSearchSelectStory />);
    },
    name: 'SingleSearchSelect',
    notes: require('./notes.md').default
  });
