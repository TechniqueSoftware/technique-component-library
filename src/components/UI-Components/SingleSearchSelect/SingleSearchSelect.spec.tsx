import * as React from 'react';
import { shallow } from 'enzyme';
import { MuiThemeProvider } from '@material-ui/core/styles';
import SingleSearchSelect, { SingleSearchSelectProps } from './SingleSearchSelect';
import clubOsTheme from '../../themes/clubOS';

describe('SingleSearchSelect', () => {
  test('should render with minimal props', () => {
    const singleSearchSelectProps: SingleSearchSelectProps = {
      id: 'single-select',
      open: true,
      label: 'test label',
      value: '1',
      options: [
        { label: 'test 1', value: '1' },
        { label: 'test 2', value: '2' },
        { label: 'test 3', value: '3' }
      ],
      disabled: false,
      maxWidth: 300,
      searchContainerMaxWidth: 300,
      listContainerHeight: 600,
      onChange: jest.fn(),
      onMenuOpen: jest.fn(),
      onMenuClose: jest.fn()
    };

    const shallowTimeWrapper = shallow(
      <MuiThemeProvider theme={clubOsTheme}>
        <SingleSearchSelect {...singleSearchSelectProps} />
      </MuiThemeProvider>
      );
    expect(shallowTimeWrapper.debug()).toMatchSnapshot();
  });
});
