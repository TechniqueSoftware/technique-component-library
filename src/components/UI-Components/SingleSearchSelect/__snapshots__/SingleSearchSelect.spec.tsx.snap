// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleSearchSelect should render with minimal props 1`] = `
"<ContextProvider value={{...}}>
  <WithStyles(SingleSearchSelect) id=\\"single-select\\" open={true} label=\\"test label\\" value=\\"1\\" options={{...}} disabled={false} maxWidth={300} searchContainerMaxWidth={300} listContainerHeight={600} onChange={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} onMenuOpen={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} onMenuClose={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} />
</ContextProvider>"
`;
